Implement feedback on data tiers (#6092)
elastic/docs-content: manage-data/lifecycle/data-tiers/manage-data-tiers-ech-ece.md

If in doubt, reach out to Support.
:::

> * Know whether you are disabling a tier that stores regular indices or {{search-snaps}}. The frozen tier only holds [partially mounted](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md#partially-mounted) {{search-snaps}}. The cold tier can hold regular indices or [fully mounted](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md#fully-mounted) {{search-snaps}}. The hot and warm tiers usually hold regular indices, but an {{ilm-init}} policy can mount a fully mounted {{search-snap}} on the hot or warm tier (for example, when the [searchable_snapshot](elasticsearch://reference/elasticsearch/index-lifecycle-actions/ilm-searchable-snapshot.md) action runs in the `hot` or `warm` phase). Use these requests to check for {{search-snap}} indices on the tier you are removing:

Reviewer: As noted top of [this doc](https://www.elastic.co/docs/reference/elasticsearch/index-lifecycle-actions/ilm-searchable-snapshot) ILM searchable snapshots only apply mounted to hot+cold and partial to frozen. Warm isn't in the mix & supposedly never hosts searchable snapshots. Info confirmation [link](https://www.elastic.co/docs/manage-data/lifecycle/index-lifecycle-management/index-lifecycle#phases-availability).

Author: Thanks, removing mentions of warm in a new commit

### Remove a tier with regular indices [non-searchable-snapshot-data-tier]

> The frozen tier only stores [partially mounted](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md#partially-mounted) {{search-snaps}}. [Fully mounted](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md#fully-mounted) {{search-snaps}} can be allocated to the hot, warm, or cold tier depending on the {{ilm-init}} phase, while the cold tier can also hold regular indices. Use the checks in [Before you remove a data tier](#before-you-remove-a-data-tier) if you are unsure what is on the tier.

Reviewer: ditto ☝️

> When you update the deployment, {{ech}} and {{ece}} try to move all data off the nodes that are removed. On {{eck}}, node set and scaling changes similarly try to relocate shards off nodes that are removed, subject to the same allocation rules, capacity, and [disk watermarks](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md#disk-based-shard-allocation) on the destination nodes. For more information, refer to [{{eck}}](/deploy-manage/deploy/cloud-on-k8s.md). To disable a tier that holds only regular indices, make sure that all data on that tier can be re-allocated by reconfiguring the relevant shard allocation filters. You’ll also need to temporarily stop your {{ilm-init}} policies to prevent new indices from being moved to the data tier you want to disable.

Reviewer 2: ```suggestion
When you update the deployment, {{ech}} and {{ece}} try to move all data off the nodes that are removed. On {{eck}}, node set and scaling changes similarly try to relocate shards off nodes that are removed, subject to the same allocation rules, capacity, and [disk watermarks](elasticsearch://reference/elasticsearch/configuration-reference/cluster-level-shard-allocation-routing-settings.md#disk-based-shard-allocation) on the destination nodes. For more information, refer to the [{{eck}}](/deploy-manage/deploy/cloud-on-k8s.md) documentation. To disable a tier that holds only regular indices, make sure that all data on that tier can be re-allocated by reconfiguring the relevant shard allocation filters. You’ll also need to temporarily stop your {{ilm-init}} policies to prevent new indices from being moved to the data tier you want to disable.
```
We might ask our friend Claude if there's a more specific section that we could link to in the ECK docs, but as is good too.

Author: I actually moved this into a note on the Configure data tiers for self-managed and Elastic Cloud on Kubernetes deployments page, since I've separated out the ECE/ECH and self-managed/ECK into child pages.

        }
        ```

>         You must remove the relevant tier from the inclusion rules. For example, to disable the warm tier, remove the `data_warm` and set `_tier_preference` to a tier you are keeping. Prefer promoting data through the lifecycle, for example from warm to cold, not back to hot, unless the cluster has no colder tier to accept the data:

Reviewer 2: ```suggestion
        You must remove the relevant tier from the inclusion rules. For example, to disable the warm tier, remove the `data_warm` parameter and set `_tier_preference` to a tier you are keeping. Prefer promoting data through the lifecycle, for example from warm to cold, not back to hot, unless the cluster has no colder tier to accept the data:
```

Author: thanks, fixed in a separate commit

        ```sh
        PUT /my-index/_settings
        {
            "routing": {
              "allocation": {
                "include": {

>                     "_tier_preference": "data_cold,data_hot" <1>

Reviewer: Per [ILM Migrate](https://www.elastic.co/docs/reference/elasticsearch/index-lifecycle-actions/ilm-migrate) the `_tier_preference` sets `data_cold,data_warm,data_hot` and not just `data_cold,data_hot`.

Author: Thanks, fixed in a separate commit

        }
        ```

>         Removing required attributes does not trigger a shard reallocation. These shards are moved when applying the plan to disable the data tier. Alternatively, you can use the [cluster re-route API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cluster-reroute) to manually re-allocate the shards before removing the nodes, or set `require` so shards target a remaining lower tier. For example, for warm to cold migration, instead of warm to hot:

Reviewer 2: ```suggestion
        Removing required attributes does not trigger a shard reallocation. These shards are moved when applying the plan to disable the data tier. Alternatively, you can use the [cluster re-route API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cluster-reroute) to manually re-allocate the shards before removing the nodes, or set `require` so shards target a remaining lower tier. For example, to set warm to cold migration, instead of warm to hot:
```

>         Removing required attributes does not trigger a shard reallocation. These shards are moved when applying the plan to disable the data tier. Alternatively, you can use the [cluster re-route API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cluster-reroute) to manually re-allocate the shards before removing the nodes, or set `require` so shards target a remaining lower tier. For example, to set warm to cold migration, instead of warm to hot:

Reviewer: ```suggestion
        Removing required attributes does not trigger a shard reallocation. These shards are moved when applying the plan to disable the data tier. Alternatively, you can use the [cluster re-route API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cluster-reroute) to manually re-allocate the shards before removing the nodes, or set `require` to migrate shards to a desired tier. For example to force an index to nodes with `data` attribute of `cold`:
```

        ```sh
        PUT /my-index/_settings
        {
            "routing": {
              "allocation": {
                "require": {
                    "data": "cold"
                }
              }
            }
        }
        ```

>         Adjust the `data` value to match the [custom node attributes](elasticsearch://reference/elasticsearch/configuration-reference/node-settings.md#custom-node-attributes) and [index-level shard allocation filters](elasticsearch://reference/elasticsearch/index-settings/shard-allocation.md) your indices already use. Prefer cold or the next lower tier over hot when you are only removing a middle tier. Do not send regular indices to the frozen tier.

Reviewer: ```suggestion
        Adjust the `data` value to match the [custom node attributes](elasticsearch://reference/elasticsearch/configuration-reference/node-settings.md#custom-node-attributes) and [index-level shard allocation filters](elasticsearch://reference/elasticsearch/index-settings/shard-allocation.md) your indices already use.  You cannot send regular indices to the frozen tier.
```

> Prefer cold or the next lower tier over hot when you are only removing a middle tier.

FWIW hot>warm>cold>frozen is industry-ish but our jargon. So there is no "next lower" other than warm between hot+cold.

It's probably not worth spending time to unwind the historical dredge here. 

But for context, historically, users used to self-manage data temperatures via the `allocation.require.data` setting. We then moved them as part of ILM over to `_tier_preference` with the understanding that they could still use the first, but it should be for quote unquote "other things" now instead of data temperatures (otherwise you can sometimes end up with blocking setting conflicts). For example, if hardware profiles are uneven and they want to self-manage that. 

But calling out in context, while definitely allowed, it's also something Support flags to users as "probably indicates your migration onto ILM is incomplete" _every time_ we see this type of `data: cold`, it's very 👻 "things are very likely going to break" flag to us.

        }
        ```

    :::{important}

>     Confirm that no shards are left on the nodes to be removed after allocation settles: `GET /_cat/shards` (filter by `node` as needed) should show the tier empty. Updating settings starts relocation, but you must wait until [shard allocation and recovery](/deploy-manage/distributed-architecture/shard-allocation-relocation-recovery.md) finish. If shards stay on the old tier, use the [cluster allocation explain](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cluster-allocation-explain) API. Common blockers can be [disk watermarks](/troubleshoot/elasticsearch/fix-watermark-errors.md) or [`index.routing.allocation.total_shards_per_node`](elasticsearch://reference/elasticsearch/index-settings/total-shards-per-node.md#total-shards-per-node) on the destination nodes.

Reviewer 2: ```suggestion
    Confirm that no shards are left on the nodes to be removed after the allocation completes: `GET /_cat/shards` (filter by `node` as needed) should show that the tier is empty. Updating settings starts the relocation process, but you must wait until [shard allocation and recovery](/deploy-manage/distributed-architecture/shard-allocation-relocation-recovery.md) finish. If shards stay on the original tier, use the [cluster allocation explain](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cluster-allocation-explain) API to determine the cause. Common reasons can be [disk watermarks](/troubleshoot/elasticsearch/fix-watermark-errors.md) or [`index.routing.allocation.total_shards_per_node`](elasticsearch://reference/elasticsearch/index-settings/total-shards-per-node.md#total-shards-per-node) on the destination nodes.
```
Maybe this improves the clarity a bit.

>     Confirm that no shards are left on the nodes to be removed after the allocation completes: `GET /_cat/shards` (filter by `node` as needed) should that show the tier is empty. Updating settings starts the relocation process, but you must wait until [shard allocation and recovery](/deploy-manage/distributed-architecture/shard-allocation-relocation-recovery.md) finish. If shards stay on the original tier, use the [cluster allocation explain](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-cluster-allocation-explain) API to determine the cause. Common reasons can be [disk watermarks](/troubleshoot/elasticsearch/fix-watermark-errors.md) or [`index.routing.allocation.total_shards_per_node`](elasticsearch://reference/elasticsearch/index-settings/total-shards-per-node.md#total-shards-per-node) on the destination nodes.

Reviewer: I like it. Noting CAT Allocation to see IF any shards remain is way easier for users to check when node gets to 0 (at least until https://github.com/elastic/elasticsearch/issues/141786). And then CAT Shards comes up to see which shards if they need to investigate.

### Remove a tier with {{search-snaps}} [searchable-snapshot-data-tier]

:::{tip}

> Fully mounted `restored-*` [{{search-snaps}}](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md#fully-mounted) on the hot, warm, or cold tier can often be moved to another remaining tier by updating [`index.routing.allocation.include._tier_preference`](#update-data-tier-allocation-rules) and related [allocation rules](/deploy-manage/distributed-architecture/shard-allocation-relocation-recovery/index-level-shard-allocation.md) alone. You do not need `POST _snapshot/.../_restore` in that case. Use a full restore to a new regular index when you are [rehydrating](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md) from a snapshot, or when you have [partially mounted](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md#partially-mounted) `partial-*` indices on frozen and need them as regular indices on another tier.

Reviewer 2: ```suggestion
Fully mounted `restored-*` [{{search-snaps}}](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md#fully-mounted) on the hot, warm, or cold tier can often be moved to another remaining tier by updating the [`index.routing.allocation.include._tier_preference`](#update-data-tier-allocation-rules) setting and related [allocation rules](/deploy-manage/distributed-architecture/shard-allocation-relocation-recovery/index-level-shard-allocation.md) alone. You do not need to run the `POST _snapshot/.../_restore` request in that case. Use a full restore to a new regular index when you are [rehydrating](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md) from a snapshot, or when you have [partially mounted](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md#partially-mounted) `partial-*` indices on a frozen tier and need them as regular indices on another tier.
```

> Fully mounted `restored-*` [{{search-snaps}}](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md#fully-mounted) on the hot, warm, or cold tier can often be moved to another remaining tier by updating the [`index.routing.allocation.include._tier_preference`](#update-data-tier-allocation-rules) setting and related [allocation rules](/deploy-manage/distributed-architecture/shard-allocation-relocation-recovery/index-level-shard-allocation.md) alone. You do not need to run `POST _snapshot/.../_restore` request in that case. Use a full restore to a new regular index when you are [rehydrating](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md) from a snapshot, or when you have [partially mounted](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md#partially-mounted) `partial-*` indices on a frozen tier and need them as regular indices on another tier.

Reviewer: Again not warm.

As right but carefulness on phrasing: users can technically self-manage searchable snapshots in which case they can call them whatever they want and them still be fully mounted. When ILM manages the index, we as Elastic chose `restored-*` as our index pattern.

Author: I've tried to reword that to make that point.

:::

> When an [{{ilm-init}} policy’s `searchable_snapshot` action](elasticsearch://reference/elasticsearch/index-lifecycle-actions/ilm-searchable-snapshot.md) runs in `hot`, `warm`, `cold`, or `frozen`, it can convert a managed index into a [{{search-snap}}](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md) (`restored-*` in non-frozen phases, or `partial-*` in `frozen`). If the data is no longer required, the delete phase of the same policy (or a manual delete) can remove it. If you must retain the data while removing the tier, follow the following steps:

Reviewer 2: ```suggestion
When an {{ilm-init}} policy’s [`searchable_snapshot` action](elasticsearch://reference/elasticsearch/index-lifecycle-actions/ilm-searchable-snapshot.md) runs in a `hot`, `warm`, `cold`, or `frozen` phase, it can convert a managed index into a [{{search-snap}}](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md) (`restored-*` in non-frozen phases, or `partial-*` in the `frozen` phase). If the data is no longer required, the `delete` phase of the same policy (or a manual delete) can remove it. If you must retain the data while removing the tier, follow these steps:
```

Author: follow the following 🤕

Author: Fixed in a new commit, thanks

> When an [{{ilm-init}} policy’s `searchable_snapshot` action](elasticsearch://reference/elasticsearch/index-lifecycle-actions/ilm-searchable-snapshot.md) runs in a `hot`, `warm`, `cold`, or `frozen` phase, it can convert a managed index into a [{{search-snap}}](/deploy-manage/tools/snapshot-and-restore/searchable-snapshots.md) (`restored-*` in non-frozen phases, or `partial-*` in the `frozen` phase). If the data is no longer required, the `delete` phase of the same policy (or a manual delete) can remove it. If you must retain the data while removing the tier, follow these steps:

Reviewer: ditto not warm

IF you just do a "manual delete" you will leave a stray searchable snapshot in the repository. If you do that in mass or over time, you can cause repository-side disk issues.

Author: ok, removing that in a separate commit

    GET _ilm/status
    ```

2. Capture a comprehensive list of index and {{search-snap}} names, and which snapshot repository each snapshot lives in.

>     1. The index name of the {{search-snaps}} may differ based on the data tier. If you intend to disable the cold tier, use the `restored-*` prefix. If the frozen tier is the one to be disabled, use the `partial-*` prefix. If you are removing a tier that had a [searchable_snapshot](elasticsearch://reference/elasticsearch/index-lifecycle-actions/ilm-searchable-snapshot.md) action in an earlier phase, for example such as hot, also run the same query for `restored-*` on that tier.

Reviewer 2: ```suggestion
    1. The index name of the {{search-snaps}} may differ based on the data tier. If you intend to disable the cold tier, use the `restored-*` prefix. If the frozen tier is the one to be disabled, use the `partial-*` prefix. If you are removing a tier that had a [searchable_snapshot](elasticsearch://reference/elasticsearch/index-lifecycle-actions/ilm-searchable-snapshot.md) action in an earlier phase, for example during the `hot` phase, also run the same query for `restored-*` on that tier.
```

        ```sh
        GET <searchable-snapshot-index-prefix>/_settings?filter_path=**.index.store.snapshot.snapshot_name&expand_wildcards=all
        ```

>     2. List snapshot repository names and resolve which repository to use in the restore API. {{ech}} often defaults to a repository named `found-snapshots`, but clusters can register additional repositories (for example with the [create snapshot repository API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-snapshot-create-repository)), so do not assume a single name.

Reviewer 2: ```suggestion
    2. List snapshot repository names to determine which repositories to use when you call the restore API. {{ech}} uses a repository named `found-snapshots` by default, but you can also register additional repositories for clusters (for example by using the [create snapshot repository API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-snapshot-create-repository)). Be careful not to assume that there is only one repository name.
```

>     2. List snapshot repository names to determine which repositories to use when you call the restore API. {{ech}} uses a repository named `found-snapshots` by default, but you can also register additional repositories for clusters (for example by using the [create snapshot repository API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-snapshot-create-repository)). Be careful not to assume that there is only one repository name.

Reviewer: The command just prior reports _per index_ both its repository name and snapshot name & since users can Searchable Snapshots point to different repositories per ILM policy (and/or its changes across versions), you don't want to run this command as a global-assume for all indices
```
filter_path=**.index.store.snapshot
```

Author: Fixing in a new commit, thanks

5. Restore indices from the {{search-snaps}}.

    1. Follow the steps to [specify the data tier based allocation inclusion rules](/manage-data/lifecycle/data-tiers/manage-data-tiers-ech-ece.md#update-data-tier-allocation-rules) for the tier you are keeping.

>     2. Use the repository and snapshot names you captured earlier in step 2. In the restore request, set `index.lifecycle.name` to `null` and set [`index.lifecycle.indexing_complete`](elasticsearch://reference/elasticsearch/configuration-reference/index-lifecycle-management-settings.md#_index_level_settings_2) to `true` for indices that are not going to keep ingesting as a rollover write index. Do not set a new `index.lifecycle.name` in the restore body. After the index is `green`, assign a different policy, if you need one, in a separate `PUT` request. Follow the steps described in [Switch lifecycle policies](/manage-data/lifecycle/index-lifecycle-management/policy-updates.md#switch-lifecycle-policies) to remove the old policy and then add a new one. Only set `index.lifecycle.rollover_alias` if you will keep using the index as a rollover target; otherwise leave it unset.

Reviewer: > if you need one, in a separate `PUT` request.

Indices restore with their earlier ILM cache so you still have to follow https://www.elastic.co/docs/manage-data/lifecycle/index-lifecycle-management/policy-updates#switch-lifecycle-policies to pivot to new policy. It's not just a "separate `PUT`" you actually have to run https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-ilm-remove-policy

Author: Ok, changing back to the original doc as I think I've made this more confusing than it was.

>     2. Use the repository and snapshot names you captured earlier in step 2. In the restore request, set `index.lifecycle.name` to `null` and set [`index.lifecycle.indexing_complete`](elasticsearch://reference/elasticsearch/configuration-reference/index-lifecycle-management-settings.md#_index_level_settings_2) to `true` for indices that are not going to keep ingesting as a rollover write index. Do not set a new `index.lifecycle.name` in the restore body. After the index is `green`, assign a different policy, if you need one, in a separate `PUT` request, following the steps described in [Switch lifecycle policies](/manage-data/lifecycle/index-lifecycle-management/policy-updates.md#switch-lifecycle-policies) which remove the old policy, then add the new one. Only set `index.lifecycle.rollover_alias` if you will keep using the index as a rollover target; otherwise leave it unset.

Reviewer 2: ```suggestion
    2. Use the repository and snapshot names you captured earlier in step 2. In the restore request, set `index.lifecycle.name` to `null` and set [`index.lifecycle.indexing_complete`](elasticsearch://reference/elasticsearch/configuration-reference/index-lifecycle-management-settings.md#_index_level_settings_2) to `true` for indices that are not going to keep ingesting as a rollover write index. Do not set a new `index.lifecycle.name` in the restore body. After the index is `green`, assign a different policy, if you need one, in a separate `PUT` request. Follow the steps described in [Switch lifecycle policies](/manage-data/lifecycle/index-lifecycle-management/policy-updates.md#switch-lifecycle-policies) to remove the old policy and then add a new one. Only set `index.lifecycle.rollover_alias` if you will keep using the index as a rollover target; otherwise leave it unset.
```

        ```sh

>         POST _snapshot/<snapshot-repository-name>/<searchable_snapshot_name>/_restore

Reviewer: ```suggestion
        POST _snapshot/<snapshot_repository_name>/<searchable_snapshot_name>/_restore
```
nitpick: formatting par to previous

        {
          "indices": "*",
          "index_settings": {
            "index.routing.allocation.include._tier_preference": "<data_tiers>",
            "index.number_of_replicas": 0,
            "index.lifecycle.name": null,
            "index.lifecycle.indexing_complete": true
          }
        }
        ```

>         The `<searchable_snapshot_name>` are obtained from step 2. Adjust `index.number_of_replicas` to match your resiliency needs.

Reviewer 2: ```suggestion
        For `<searchable_snapshot_name>` use the names that you obtained in step 2. Adjust `index.number_of_replicas` to match your resiliency needs.
```
